security/pkg/nodeagent/util/util.go

```diff
@@ -53,3 +55,24 @@ func GetMetricsCounterValue(metricName string) (float64, error) {
 
 	return rows[0].Data.(*view.SumData).Value, nil
 }
+
+// Output the key and certificate to the given directory.
+// If directory is empty, return nil.
+func OutputKeyCertToDir(dir string, priviateKey, certChain, rootCert []byte) error {
+	if len(dir) == 0 {
+		return nil
+	}
+	// Depending on the SDS resource to output, some fields may be nil
+	if priviateKey == nil && certChain == nil && rootCert == nil {
+		return fmt.Errorf("the input private key, cert chain, and root cert are nil")
+	}
+	err1 := ioutil.WriteFile(path.Join(dir, "key.pem"), priviateKey, 0777)
+	err2 := ioutil.WriteFile(path.Join(dir, "cert-chain.pem"), certChain, 0777)
+	err3 := ioutil.WriteFile(path.Join(dir, "root-cert.pem"), rootCert, 0777)
+
+	if err1 != nil && err2 != nil && err3 != nil {
```

Comment:
shouldn't these be `||`?

Reply:
Depending on the SDS resources, some of the fields to write may be nil and trigger a write error. Hence the || instead of &&.

Reply:
But you have && here so all 3 need to fail in order for this code path to be hit?

Reply:
Yes, this code is only hit when all three fail. The reason is that for some SDS resource (e.g., ROOTCA), only one field is non-nil.

Reply:
I think you are mixing the nil data error and IO error? Can you do:
```
if privateKey != nil && err := ioutil.WriteFile(path.Join(dir, "key.pem"), priviateKey, 0777), err != nil {
  return fmt.Errorf("Failed to write private key to file: ", err)
}
if certChain != nil && err := ioutil.WriteFile(path.Join(dir, "cert-chain.pem"), certChain, 0777), err != nil {
  return fmt.Errorf("Failed to write cert chain to file: ", err)
}
if rootCert != nil && err := ioutil.WriteFile(path.Join(dir, "root-cert.pem"), rootCert, 0777), err != nil {
  return fmt.Errorf("Failed to write root cert to file: ", err)
}
```

Reply:
@myidpt Done, thanks.